nibabel/nifti1.py

```diff
@@ -698,16 +698,26 @@ def get_data_shape(self):
         Allows for freesurfer hack for large vectors described in
         https://github.com/nipy/nibabel/issues/100 and
         http://code.google.com/p/fieldtrip/source/browse/trunk/external/freesurfer/save_nifti.m?spec=svn5022&r=5022#77
+
+        Allows for freesurfer hack for 7th order icosahedron surface described
+        in
+        https://github.com/nipy/nibabel/issues/309
+        https://code.google.com/p/fieldtrip/source/browse/trunk/external/freesurfer/load_nifti.m?r=8776#86
+        https://code.google.com/p/fieldtrip/source/browse/trunk/external/freesurfer/save_nifti.m?r=8776#50
         '''
         shape = super(Nifti1Header, self).get_data_shape()
         # Apply freesurfer hack for vector
-        if shape != (-1, 1, 1): # Normal case
+        if shape == (-1, 1, 1):
+            vec_len = int(self._structarr['glmin'])
+            if vec_len == 0:
+                raise HeaderDataError('-1 in dim[1] but 0 in glmin; '
+                                      'inconsistent freesurfer type header?')
+            return (vec_len, 1, 1)
+        # Apply freesurfer hack for ico7 surface
+        elif shape == (27307, 1, 6):
```

Comment:
The easiest way to clarify the situation is to simply try all of the icosahedra:

    $ for ORDER in {1..7}; do
        mri_surf2surf --hemi rh --srcsubject fsaverage \
            --srcsurfval $SUBJECTS_DIR/fsaverage/surf/rh.orig.avg.area.mgh \
            --trgsubject ico --trgsurfval rh.ico$ORDER.nii \
            --trgicoorder $ORDER &> /dev/null \
        && python -c "from nibabel import load;\
              print ($ORDER, load('rh.ico$ORDER.nii').header._structarr['dim'][1:4])" \
        && rm rh.ico$ORDER.nii
      done 
    (1, array([42,  1,  1], dtype=int16))
    (2, array([162,   1,   1], dtype=int16))
    (3, array([642,   1,   1], dtype=int16))
    (4, array([2562,    1,    1], dtype=int16))
    (5, array([10242,     1,     1], dtype=int16))
    (6, array([-1,  1,  1], dtype=int16))
    (7, array([27307,     1,     6], dtype=int16))

So whatever their ideal of what ought to be done for icosahedra with >2^15 vertices, in practice, they use the (27307, 1, 6) shape for ico7, the large vector hack for ico6 and just (x, 1, 1) for lesser icosahedra. (ico8 does not work.)